### PR TITLE
[TS migration] Migrate 'OptionRow.js' component to TypeScript

### PR DESCRIPTION
**Pull request by VickyStash**
### Details

[TS migration] Migrate 'OptionRow.js' component to TypeScript

### Fixed Issues

$ https://github.com/Expensify/App/issues/25118
PROPOSAL: N/A

### Tests

- [x] Verify that no errors appear in the JS console

**First case:**
1. Click FAB and select `Start chat`
2. Make sure the list of users, names, and avatars look as expected.
3. [WEB AND DESKTOP] Press Arrow Up/Down and verify that the focused state changes.
4. [WEB AND DESKTOP] Verify the hover effect over the Option Row. When hovering over `Add to group` button, the hover state should be removed.
5. Toggle select status by clicking `Add to group `or checkbox button and verify that the button is changed correctly.

**Second case:**
1. Click FAB and select `Request money`
2. Go to `Manual` section, enter any amount of money and press Next. You should see the list of users and workspaces.
3. Make sure the list looks as expected, all names and avatars look correct.
4. [WEB AND DESKTOP] Press Arrow Up/Down and verify that the focused state changes.
5. [WEB AND DESKTOP] Verify the hover effect over the Option Row. When hovering over `Split` button, the hover state should be removed.
6. Toggle select status by clicking `Split `or checkbox button and verify that the button is changed correctly.

**Third case:**
1. Open any room.
2. Press on the room header to open room details.
3. Click on members.
4. Make sure the members list look as expected.
5.  [WEB AND DESKTOP]  Verify the hover effect over the Option Row works as expected, and the tooltip appears if you hover over the user name.

**Storybook:**
1. Open Storybook
6. Select `OptionRow` component and make sure it looks ok.

### Offline tests

N/A

### QA Steps

- [x] Verify that no errors appear in the JS console

Same as in the **Tests** section

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/23176449/72827ade-f451-4dd2-b67e-dfe82a7fee92

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/23176449/2edc8cec-7660-47e1-b7cc-60962e68bba4

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/23176449/65ff37e8-5ab5-4fbe-9303-c3e74deece32

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/23176449/f0683c24-3c5c-4456-8d4c-9b29d536bedf

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/23176449/ad531047-d6e3-4e4d-b93e-ecb148061951

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/23176449/ee6e03ea-dc4a-411a-97ec-ea71803a4bc8

</details>

<details>
<summary>Storybook</summary>

storybook

</details>

### CODE REVIEW

File: src/components/OptionRow.tsx [hunk 1]
```diff
@@ -16,169 +16,164 @@ import Icon from './Icon';
...
+function OptionRow({
+    option,
+    onSelectRow,
+    style,
+    hoverStyle,
+    isDisabled: isOptionDisabled = false,
+    isMultilineSupported = false,
+    shouldShowSelectedStateAsButton = false,
+    highlightSelected = false,
+    selectedStateButtonText = 'Select',
```
Comment (fabioh8010) [inline src/components/OptionRow.tsx:90]: Shouldn't this be in the translation file? 🤔

File: src/libs/ReportUtils.ts [hunk 2]
```diff
@@ -317,9 +318,17 @@ type OptionData = {
     allReportErrors?: Errors | null;
     brickRoadIndicator?: typeof CONST.BRICK_ROAD_INDICATOR_STATUS.ERROR | '' | null;
     tooltipText?: string | null;
+    alternateTextMaxLines?: number;
+    boldStyle?: boolean;
+    customIcon?: {
+        src: (props: SrcProps) => React.ReactNode;
+        color?: string;
+    };
```
Comment (fabioh8010) [inline src/libs/ReportUtils.ts:326]: Could we extract this into a separate type?

File: src/styles/utils/index.ts [hunk 3]
```diff
@@ -383,7 +383,11 @@ function getWidthStyle(width: number): ViewStyle {
 /**
  * Returns a style with backgroundColor and borderColor set to the same color
  */
-function getBackgroundAndBorderStyle(backgroundColor: string): ViewStyle {
+function getBackgroundAndBorderStyle(backgroundColor: string | undefined): ViewStyle {
+    if (!backgroundColor) {
+        return {};
+    }
```
Comment (fabioh8010) [inline src/styles/utils/index.ts:389]: What is the reason of this change?

File: src/components/OptionRow.tsx (hunk 1 above)
Comment (VickyStash) [reply src/components/OptionRow.tsx:90]: @fabioh8010 I've not found it there... Since `selectedStateButtonText` is used only when `shouldShowSelectedStateAsButton` is `true`, and in the default values it's `false` maybe we can turn the default value into an empty string.


Another suggestion is to remove the default value at all, add _Select_ into translation vocabularies and update the code to be
```
text={selectedStateButtonText ?? translate('common.select')}
```

Comment (fabioh8010) [reply src/components/OptionRow.tsx:90]: I think is best to add this string to the translations!

File: src/components/Icon/index.tsx [hunk 4]
```diff
@@ -103,3 +103,4 @@ class Icon extends PureComponent<IconProps> {
 }
 
 export default withTheme(withThemeStyles(withStyleUtils(Icon)));
+export type {SrcProps};
```
Comment (pasyukevich) [inline src/components/Icon/index.tsx:106]: Can we move this type to the `type.ts`?

Comment (VickyStash) [reply src/components/Icon/index.tsx:106]: @pasyukevich Which one? Or do you mean to create a new one? Usually, we create` types.ts` to share the types between multi-platform implementation

**Review by fabioh8010 — APPROVED** [review]

File: src/components/Icon/index.tsx (hunk 4 above)
Comment (pasyukevich) [reply src/components/Icon/index.tsx:106]: Yeah, I mean to create a new one, since this type is used outside also

Comment (VickyStash) [reply src/components/Icon/index.tsx:106]: Since we have other places with the same approach, I think we can leave it as is, at least for now.

File: src/components/Pressable/GenericPressable/types.ts [hunk 5]
```diff
@@ -19,7 +19,7 @@ type RequiredAccessibilityLabel =
            * @example 'Search'
            * @example 'Close'
            */
-          accessibilityLabel: string;
+          accessibilityLabel?: string;
```
Comment (blazejkustra) [inline src/components/Pressable/GenericPressable/types.ts:22]: I think this should be a required prop @VickyStash

File: src/components/OptionRow.tsx [hunk 6]
```diff
@@ -288,15 +281,15 @@ function OptionRow(props) {
...
                             <View
                                 style={[styles.flexRow, styles.alignItemsCenter]}
                                 accessible={false}
                             >
                                 <View>
                                     <Icon
-                                        src={lodashGet(props.option, 'customIcon.src', '')}
-                                        fill={lodashGet(props.option, 'customIcon.color')}
+                                        src={option.customIcon.src}
+                                        fill={option.customIcon.color}
```
Comment (blazejkustra) [inline src/components/OptionRow.tsx:292]:
```suggestion
                                        src={option.customIcon?.src}
                                        fill={option.customIcon?.color}
```

**Review by blazejkustra — APPROVED** [review]

File: src/components/OptionRow.tsx (hunk 6 above)
Comment (VickyStash) [reply src/components/OptionRow.tsx:292]: @blazejkustra It's checked in [this line](https://github.com/Expensify/App/pull/33077/files/2def510e3683e5197cc3e75f657b6d3ff5596f8d#diff-e7268ea8328cb44bb4424e15bdd5559fb0655b1b83eb1f5a88ac4a2683966d14R284)

File: src/components/Pressable/GenericPressable/types.ts (hunk 5 above)
Comment (VickyStash) [reply src/components/Pressable/GenericPressable/types.ts]: @blazejkustra Updated, please take a look

File: src/components/OptionRow.tsx [hunk 7]
```diff
@@ -326,7 +318,7 @@ export default React.memo(
         prevProps.showSelectedState === nextProps.showSelectedState &&
         prevProps.highlightSelected === nextProps.highlightSelected &&
         prevProps.showTitleTooltip === nextProps.showTitleTooltip &&
-        !_.isEqual(prevProps.option.icons, nextProps.option.icons) &&
+        !lodashIsEqual(prevProps.option.icons, nextProps.option.icons) &&
```
Comment (situchan) [inline src/components/OptionRow.tsx:321]: Please fix conflict.
`!lodashIsEqual` > `lodashIsEqual` to fix https://github.com/Expensify/App/issues/33353

Comment (VickyStash) [reply src/components/OptionRow.tsx]: @situchan Done, thank you!

**Comment by situchan** [comment]
Seems like recent regression - default avatar not showing at all on native. Also happening on main so out of scope

Screenshot 2023-12-29 at 1 37 55 PM

**Comment by situchan** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/108292595/35305bfb-a07f-4299-9b2a-a6266294e39c




</details>

<details>
<summary>Android: mWeb Chrome</summary>


https://github.com/Expensify/App/assets/108292595/50dc5a78-4951-47e2-88db-7a07a3158b32



</details>

<details>
<summary>iOS: Native</summary>


https://github.com/Expensify/App/assets/108292595/8a8e3e17-b4fd-4fbc-88b0-938c83d825cd



</details>

<details>
<summary>iOS: mWeb Safari</summary>



https://github.com/Expensify/App/assets/108292595/31d643fc-b625-41a4-a62d-61a379eab23e


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/Expensify/App/assets/108292595/b0b80d30-a24e-4e71-80eb-cdd5e5c8cc74



</details>

<details>
<summary>MacOS: Desktop</summary>


https://github.com/Expensify/App/assets/108292595/08158bdf-b081-4441-984c-9de96cf5d791



</details>

**Review by situchan — APPROVED** [review]
:shipit:

**Review by situchan — APPROVED** [review]

**Review by mountiny — APPROVED** [review]
No comments for the code, this is highly used component so please be on a look out for any regressions cc @situchan
